src/scr_flush_async.c

```diff
@@ -9,13 +9,893 @@
+                                 struct cppr_op_info *handles_ptr,
+                                 const int length)
+{
+  struct scr_cppr_info *tmp_ptr = NULL;
+  int i;
+  if(metadata_ptr != NULL){
+    for(i=0; i < length; i++){
+      tmp_ptr = &metadata_ptr[i];
+      if(tmp_ptr->alloced == true){
+        free(tmp_ptr->src_dir);
+        free(tmp_ptr->dst_dir);
+        free(tmp_ptr->filename);
+      }
+    }
+    free(metadata_ptr);
+    metadata_ptr = NULL;
+  }
+  if(handles_ptr != NULL){
+    free(handles_ptr);
+    handles_ptr = NULL;
+  }
+}
+#endif /* #ifdef HAVE_LIBCPPR */
+
+/*
+=========================================
+Asynchronous flush helper functions
+========================================
+*/
+
+/* dequeues files listed in hash2 from hash1 */
+static int scr_flush_async_file_dequeue(scr_hash* hash1, scr_hash* hash2)
+{
+  /* for each file listed in hash2, remove it from hash1 */
+  scr_hash* file_hash = scr_hash_get(hash2, SCR_TRANSFER_KEY_FILES);
+  if (file_hash != NULL) {
+    scr_hash_elem* elem;
+    for (elem = scr_hash_elem_first(file_hash);
+         elem != NULL;
+         elem = scr_hash_elem_next(elem))
+    {
+      /* get the filename, and dequeue it */
+      char* file = scr_hash_elem_key(elem);
+      scr_hash_unset_kv(hash1, SCR_TRANSFER_KEY_FILES, file);
+    }
+  }
+  return SCR_SUCCESS;
+}
+
+
+/*
+=========================================
+CPPR Asynchronous flush wrapper functions
+========================================
+*/
+#ifdef HAVE_LIBCPPR
+
+/* check whether the flush from cache to parallel file system has completed */
+static int scr_cppr_flush_async_test(scr_filemap* map, int id, double* bytes)
+{
+/* CPPR: essentially test the grp
+ * cppr_return_t cppr_test_all(uint32_t count, struct cppr_op_info info[]);
+ * each rank 0 on a node needs to call test_all, then report that in
+ * transfer_complete, then call scr_alltrue(transfer_complete)
+ * make sure transfer_complete is set to 1 in all non 0 ranks
+ */
+  /* assume the transfer is complete */
+  int transfer_complete = 1;
+
+  /* current operation for easier access */
+  struct scr_cppr_info *current_cppr_metadata = NULL;
+
+  /* current handle for easier access */
+  struct cppr_op_info *current_cppr_handle = NULL;
+
+  cppr_return_t cppr_retval;
+
+  /* initialize bytes to 0 */
+  *bytes = 0.0;
+
+  /* if user has disabled flush, return failure */
+  if (scr_flush <= 0) {
+    return SCR_FAILURE;
+  }
+  scr_dbg(1,"scr_cppr_flush_async_test called @ %s:%d\n",
+          __FILE__, __LINE__);
+
+
+  /* have master on each node check whether the flush is complete */
+  double bytes_written = 0.0;
+  if (scr_storedesc_cntl->rank == 0) {
+
+    int i;
+    cppr_retval = cppr_test_all(scr_flush_async_cppr_index + 1, cppr_ops);
+    if( cppr_retval != CPPR_SUCCESS){
+      scr_dbg(0, "CPPR ERROR WITH initial call to cppr_test(): %d",
+              cppr_retval);
+      return SCR_FAILURE;
```

Comment:
Does cppr_test_all return the same value on all procs?  If not, a return here could lead to a deadlock in the allreduce later on.